### PR TITLE
Parsing scalar as folded or literal style

### PR DESCRIPTION
**Pull request by lwronski**
Add support for parsing scalar as folded or literal style

### CODE REVIEW

File: src/main/scala/org/virtuslab/internal/load/reader/ReaderCtx.scala [hunk 1]
```diff
@@ -61,7 +58,17 @@ case class ReaderCtx(
       case Some(ReaderState.Mapping(i)) if i == indent => true
       case _                                           => false
 
-  private def closeOpenedScopes(): List[Token] =
+  def isAllowedSpecialCharacter(char: Char): Boolean =
+    stateStack.peek() match
+      case Some(ReaderState.FlowMapping) if char == '}' => false
+      case _                                            => true
```
Comment (kpodsiad) [inline src/main/scala/org/virtuslab/internal/load/reader/ReaderCtx.scala:64]: Could you explain why `FlowMapping` + `}` is illegal?

File: src/main/scala/org/virtuslab/internal/load/reader/YamlReader.scala [hunk 2]
```diff
@@ -39,35 +39,37 @@ class YamlReader(in: CharSequence) extends Reader {
       case Some('.') if isDocumentEnd    => parseDocumentEnd()
       case Some('[') =>
         skipCharacter()
-        List(ctx.appendSequence(indent))
+        ctx.appendState(ReaderState.Sequence(indent))
+        List(SequenceStart)
       case Some(']') =>
         skipCharacter()
-        List(ctx.popTokenFromStack)
+        ctx.closeOpenningSequence()
```
Comment (kpodsiad) [inline src/main/scala/org/virtuslab/internal/load/reader/YamlReader.scala:46]: Maybe `closeOpenedSequene`?

File: src/main/scala/org/virtuslab/internal/load/reader/YamlReader.scala [hunk 3]
```diff
@@ -83,9 +85,45 @@ class YamlReader(in: CharSequence) extends Reader {
...
     val scalar = readScalar()
     List(Scalar(scalar, ScalarStyle.DoubleQuoted))
-  }
 
-  private def fetchFoldedValue(): List[Token] = {
+  private def parseBlockHeader(): Unit =
+    while (peek() == Some(' '))
+      skipCharacter()
+
+    if peek() == Some('\n') then skipCharacter()
```
Comment (kpodsiad) [inline src/main/scala/org/virtuslab/internal/load/reader/YamlReader.scala:93]: Is block header meaningful name from yaml spec? I am afraid that this name is a little bit enigmatic :/

File: src/main/scala/org/virtuslab/internal/load/reader/ReaderCtx.scala (hunk 1 above)
Comment (lwronski) [reply src/main/scala/org/virtuslab/internal/load/reader/ReaderCtx.scala:64]: From yaml specifications:

>  Flow mappings are denoted by surrounding “{” and “}” characters.

so, It's not allowed, that scalar can contain special characters '}'.  The special indicator '}' is reserved for FlowMapping.

File: src/main/scala/org/virtuslab/internal/load/reader/YamlReader.scala (hunk 3 above)
Comment (lwronski) [reply src/main/scala/org/virtuslab/internal/load/reader/YamlReader.scala:93]: Yes, exactly, this name is from yaml spec - [block-header](https://yaml.org/spec/1.2/spec.html#c-b-block-header(m,t))

File: src/main/scala/org/virtuslab/internal/load/reader/YamlReader.scala [hunk 4]
```diff
@@ -94,34 +132,49 @@ class YamlReader(in: CharSequence) extends Reader {
...
+    parseBlockHeader()
+    skipUntilNextIndent(foldedIndent)
+
+    def chompedEmptyLines() =
+      while (peekNext() == Some('\n')) {
+        skipCharacter()
+        sb.append("\\n")
+      }
+
+      skipCharacter()
```
Comment (kpodsiad) [inline src/main/scala/org/virtuslab/internal/load/reader/YamlReader.scala:145]: Why additional `skipCharacter`? `skipUntilNextIndent` isn't enough?

File: src/main/scala/org/virtuslab/internal/load/reader/YamlReader.scala [hunk 5]
```diff
@@ -190,14 +247,15 @@ class YamlReader(in: CharSequence) extends Reader {
       case Some('"')  => fetchDoubleQuoteValue()
       case Some('\'') => fetchSingleQuoteValue()
       case Some('>')  => fetchFoldedValue()
+      case Some('|')  => parseLiteral()
       case _          => parseScalarValue()
```
Comment (kpodsiad) [inline src/main/scala/org/virtuslab/internal/load/reader/YamlReader.scala:251]: We shouldn't mix `fetch` and `parse`. What do you say about `parse`?

File: src/main/scala/org/virtuslab/internal/load/reader/ReaderCtx.scala [hunk 6]
```diff
@@ -36,20 +36,17 @@ case class ReaderCtx(
...
 
-  def popTokenFromStack: Token =
+  def closeOpenningSequence(): List[Token] =
     stateStack.pop() match
-      case Some(ReaderState.Sequence(_)) => Token.SequenceEnd
-      case Some(ReaderState.Mapping(_))  => Token.MappingEnd
-      case Some(ReaderState.Document)    => popTokenFromStack
-      case _                             => Token.StreamEnd
+      case Some(ReaderState.Sequence(_)) => List(Token.SequenceEnd)
+      case _                             => Nil
```
Comment (kpodsiad) [inline src/main/scala/org/virtuslab/internal/load/reader/ReaderCtx.scala:49]: It seems that `closeX` and `stateStack.pop()` could use some error handling in future iterations :D

**Review by kpodsiad — APPROVED** [review]
LGTM, but I have to look once more at `ParserImpl`

File: src/main/scala/org/virtuslab/internal/load/reader/ReaderCtx.scala (hunk 6 above)
Comment (lwronski) [reply src/main/scala/org/virtuslab/internal/load/reader/ReaderCtx.scala:49]: Good point, it should return error from input tokenising

File: src/main/scala/org/virtuslab/internal/load/reader/YamlReader.scala (hunk 4 above)
Comment (lwronski) [reply src/main/scala/org/virtuslab/internal/load/reader/YamlReader.scala:145]: No, we have to also skip new line.

**Pull request merged**